### PR TITLE
[Model][Generative Recommendation] Add HSTU model support to vLLM

### PR DESCRIPTION
**Pull request by Trenx-J**
## Purpose

This PR adds HSTU-based generative recommendation model support to vLLM.

HSTU is a ranking model architecture designed for generative recommendation. Unlike traditional language models, HSTU does not autoregressively generate text tokens. Instead, it synchronously computes ranking scores for all candidate items in a request.

The goal of this change is to enable generative recommendation models to reuse the existing vLLM and large-model serving ecosystem, including:

- Hugging Face model and configuration loading.
- vLLM model registration.
- GPU model execution.
- Attention backend selection.
- Optimized CUDA operators.
- Request batching and scheduling.
- Asynchronous output handling.
- CUDA Graph execution.
- Future distributed inference and online serving infrastructure.

Generative recommendation and large language models have several common characteristics. Both can be represented as sequence models, use embedding layers and transformer-style computation, and benefit from optimized attention kernels, GPU batching, asynchronous execution, and production-level model serving infrastructure.

By integrating HSTU into vLLM, generative recommendation models can reuse the existing large-model ecosystem instead of maintaining an independent inference stack.

## Inference Paradigm

Although HSTU and large language models share several model and runtime components, their inference semantics are different.

| Aspect | Large language model | Generative recommendation model |
| --- | --- | --- |
| Decode behavior | Autoregressively generates one token at a time | Synchronously computes scores for all candidates in a request |
| Model output | Vocabulary logits for the next token | One ranking score for each candidate item |
| Sampling | Requires token sampling and token ID generation | Does not generate a sampled token |
| Request progress | Continues until the requested number of tokens is generated | Completes after all candidate scores are returned |
| Output format | Usually `[num_tokens, vocab_size]` | Packed candidate-score output |
| Scheduler signal | Uses sampled token IDs to update request state | Uses candidate scores to complete the request |

The default vLLM generation path is designed around language-model execution:

1. Run the model forward pass.
2. Produce vocabulary logits.
3. Invoke the token sampler.
4. Generate token IDs.
5. Update the request state.
6. Continue decoding until the request is finished.

This flow cannot be directly reused for HSTU because HSTU produces candidate ranking scores rather than vocabulary logits.

Passing HSTU outputs through the normal sampler would incorrectly interpret candidate scores as vocabulary logits and could produce invalid token IDs.

Therefore, this PR adds a score-only execution path:

1. Run the HSTU forward pass.
2. Extract the score corresponding to each candidate item.
3. Convert candidate IDs and scores into `LogprobsTensors`.
4. Return the scores through `prompt_logprobs_dict`.
5. Skip the standard token sampler.
6. Skip generated-token state updates.
7. Complete the request after all candidate scores are returned.

The scheduler is also updated to support score-only recommendation requests. A request can be completed after candidate scores are produced even when no `new_token_ids` are returned.

## Implementation Overview

This PR contains five main categories of changes.

### 1. Model Structure and Model Registration

Files:

- `vllm/model_executor/models/hstu.py`
- `vllm/model_executor/models/registry.py`
- `tests/models/registry.py`

Changes:

- Add the HSTU model implementation.
- Add HSTU embedding layers, transformer blocks, normalization layers, and prediction heads.
- Register the HSTU model with vLLM.
- Support Hugging Face HSTU model loading.
- Add HSTU to the model registry test data.
- Add model registration and model import coverage.

### 2. Model Configuration and Configuration Registration

Files:

- `vllm/transformers_utils/config.py`
- `vllm/transformers_utils/configs/__init__.py`
- `vllm/transformers_utils/configs/hstu.py`

Changes:

- Add HSTU-specific configuration classes.
- Support HSTU model and task configuration fields.
- Support `model_type = "hstu"`.
- Convert HSTU configuration data from Hugging Face configuration files into the vLLM configuration representation.
- Preserve compatibility with the existing vLLM configuration loading flow.

### 3. Model Operators and Operator Registration

Files:

- `vllm/v1/attention/backends/hstu_attn.py`
- `vllm/v1/attention/backends/registry.py`
- `vllm/v1/attention/selector.py`

Changes:

- Add the HSTU attention backend.
- Integrate the external HSTU variable-length attention operator.
- Register the HSTU attention backend with vLLM.
- Automatically select `HSTU_ATTN` for HSTU or generative recommendation models when no explicit backend is configured.
- Preserve explicit attention backend configuration as the highest-priority option.
- Validate unsupported attention combinations and return clear errors.
- Pass per-request context and candidate metadata to the HSTU attention implementation.

Automatic backend selection is only applied when the user has not explicitly selected an attention backend. This preserves the existing backend override behavior.

### 4. Model Runner V1 Adaptation

File:

- `vllm/v1/worker/gpu_model_runner.py`

Changes:

- Add generative recommendation metadata collection.
- Add HSTU candidate score extraction.
- Pass HSTU metadata through the forward context.
- Bypass the standard token sampler for HSTU requests.
- Return candidate scores through the prompt-logprobs output path.
- Add recommendation-specific request completion handling.
- Preserve the existing execution path for non-HSTU models.

### 5. Model Runner V2 Adaptation

Files:

- `vllm/v1/worker/gpu/model_runner.py`
- `vllm/v1/worker/gpu/gr_state.py`
- `vllm/v1/worker/gpu/async_utils.py`
- `vllm/v1/core/sched/scheduler.py`

Changes:

- Add `GRState` for request-scoped generative recommendation state.
- Store user IDs, candidate count, prompt token IDs, and request stage.
- Move candidate score extraction from the Model Runner into `GRState`.
- Convert packed HSTU outputs into `LogprobsTensors`.
- Add asynchronous score output handling.
- Return candidate scores through `prompt_logprobs_dict`.
- Skip token sampling and token-state updates for HSTU requests.
- Complete score-only requests in the scheduler when candidate scores are available.
- Preserve the one-token early-return behavior required by generative recommendation inputs.

### 1. Environment Setup
The operator is installed inside the following NVIDIA PyTorch container:`nvcr.io/nvidia/pytorch:26.06-py3`
Install the required system dependency:`apt update && apt install -y libxcb1`

### 2.HSTU Operator Installation and Validation
The HSTU attention operator is provided by the NVIDIA `recsys-examples` repository.
Clone the recsys-examples repository:`git clone https://github.com/NVIDIA/recsys-examples.git && cd recsys-examples/corelib/hstu`
For H20 and other Hopper-based devices, enter the corresponding device directory before compiling: `cd hopper`
Install the HSTU operator:`Install the HSTU operator:`
After the compilation completes successfully, validate the operator installation with: `python vllm/tests/models/generativemodel/test_hstu_op.py`

### 3.vLLM Installation
Clone the vLLM repository:`git clone https://github.com/vllm-project/vllm.git && cd vllm`
Install the current vLLM source with CUDA support:`VLLM_TARGET_DEVICE=cuda pip install -v -e .`

### 4. HSTU Model Inference:
After the HSTU changes are included in vLLM, run:`python vllm/tests/models/generativemodel/test_hstu.py`

This test validates:
* HSTU model loading.
* HSTU configuration loading.
* HSTU model registration.
* HSTU attention backend initialization.
* HSTU forward execution.
* Candidate score extraction.
* Candidate ID and score formatting.
* Score-only execution without token sampling.
* Recommendation request completion.
* Model Runner integration.
* CUDA Graph execution.

### 5. Model Registry and Hugging Face Loading
Run the model registry tests:`pytest -q tests/models/test_registry.py -k HSTUForCausalLM`
This validates that:
* HSTUForCausalLM is registered.
* The HSTU model class can be imported.
* Hugging Face model metadata is available.
* The HSTU configuration can be loaded.
* The vLLM model registry can resolve the HSTU model.

### HSTU Operator
The test returned like:
```text
Input q shape: torch.Size([40, 4096])
Output shape: torch.Size([40, 4096])
Run success
```

### HSTU Model Inference
The test returned candidate IDs and their corresponding ranking scores. A truncated example of the output is:
```text
Request 0:
  12676:0.6602
  47052:0.5078
  45082:0.4336
  79131:0.3984
  34671:0.4844
  5695:0.4570
  95647:0.1777
  60217:0.5273
  70284:0.4258
  16361:0.4238

Request 1:
  8023:0.4863
  24460:0.6094
  ...
```
Each line follows the format:`Each line follows the format:`
The output confirms that HSTU synchronously computes the scores of all candidates for a request and returns the scores through the generative recommendation output path.
The test also confirms that:
* HSTU outputs are not passed through the standard vocabulary sampler.
* No artificial generated token IDs are produced.
* Requests complete after candidate scores are returned.
* Candidate scores are returned through prompt_logprobs_dict.
* Both Model Runner V1 and Model Runner V2 execution paths are supported.
* The HSTU attention backend is initialized and executed successfully.
* CUDA Graph execution works in the validated CUDA environment.

### Model Registration and Hugging Face Loading
The model registration and Hugging Face loading tests passed.
HSTUForCausalLM can be resolved and imported through the vLLM model registry, and the HSTU configuration can be loaded from the Hugging Face model configuration.

### Scheduler and Score-Only Execution
The scheduler validation passed for score-only recommendation requests.
A recommendation request is correctly completed after candidate scores are produced, even when no new_token_ids are returned.
The standard token sampling path is skipped for HSTU requests, and no regular language-model token state update is performed.
### Attention Backend Selection
When no explicit attention backend is configured, the HSTU model automatically selects HSTU_ATTN.
When an explicit attention backend is configured, the user-provided backend remains authoritative and is not overridden by the HSTU automatic selection logic.

### Current Scope and Limitations
This initial integration focuses on bringing HSTU into the vLLM model and serving ecosystem.
The current implementation assumes:
* Prompt lengths are consistent across users.
* Candidate lengths are consistent across users.
* HSTU and non-HSTU requests are not mixed in the same batch.
* Candidate scores are produced during an unchunked prefill.
* Pipeline parallelism is not used.
* PD-separated HSTU execution is not supported yet.
* KV-cache-based or paged HSTU attention is not supported yet.
* request_stage = 1 is not supported.
* Each recommendation prompt contains at least one context/candidate pair.
These limitations can be addressed in follow-up work after the basic HSTU execution path and serving interface are established.

### CODE REVIEW

**Comment by Trenx-J** [comment]
@hmellor Hello, our team is in charge of the generative recommendation model ecosystem. We aim to integrate generative recommendation models into the LLM ecosystem based on vLLM. Could you please review this PR and evaluate the requirements for merging this feature and corresponding models? Thank you very much for your support.

**Comment by DarkLight1337** [comment]
Do you have any example HF Hub repos that use this architecture?

**Comment by Trenx-J** [comment]
> Do you have any example HF Hub repos that use this architecture?

Yes, we have already uploaded the configs of our test models to Hugging Face. Both Lingqu‑Rec/HSTU‑1B and Lingqu‑Rec/HSTU‑0.2B are ready for use.

**Comment by Trenx-J** [comment]
> Do you have any example HF Hub repos that use this architecture?

We have also uploaded the test script for the model at tests/models/generativemodel/test_hstu.py. You can directly run this file with Python. It loads models from HF Hub repos.
The relevant model‑loading snippet is shown below:
```python
  llm = LLM(
          model="Lingqu-Rec/HSTU-0.2B",
          skip_tokenizer_init=True,
          trust_remote_code=True,
          max_num_batched_tokens=20480,
          compilation_config={ 
              'cudagraph_mode': 'FULL_DECODE_ONLY',
              'cudagraph_capture_sizes': [(j + 1) * 128 for j in range(32)] + [2, 4, 8, 16, 32, 64]
          },
      )
```

**Comment by DarkLight1337** [comment]
As this PR touches engine core, I prefer keeping this as a vLLM fork rather than having to maintain it upstream, until the model gets more popular.

**Comment by Trenx-J** [comment]
> As this PR touches engine core, I prefer keeping this as a vLLM fork rather than having to maintain it upstream, until the model gets more popular.

Thank you for your feedback. We fully understand your concern about modifications to the vLLM engine core. I would like to share our considerations below:

First, generative recommendation models are rapidly evolving. Starting from 2022, the number of open‑source generative recommendation models has grown from over 20 to more than 100, including RankMixer, OneRec, OneTrans and others. Model sizes also keep expanding, covering 0.01B, 1B, 7B, 15B up to 30B. We intend to drive support for more generative recommendation models down the line, and eventually build a standalone feature module similar to multimodal in vLLM.

Second, compared with multimodal models, generative recommendation models require relatively limited code changes within vLLM. Our team has done internal evaluation: only a small set of modifications are needed to enable generative recommendation support. Therefore we do not foresee the need for a separate derivative project like vllm‑omni.

Third, the HSTU model originates from NVIDIA’s open‑source recsys‑example repository. Given vLLM’s mature and robust KV cache management subsystem, we hope this feature can be upstreamed into the official vLLM codebase.

Fourth, our modifications touching the engine core mainly adapt the input‑output paradigm for generative recommendation workloads. The current implementation has already taken other generative recommendation model scenarios into account, so heavy follow‑up changes will not be required. If you spot any concerns regarding code structure or logic, we are happy to further refactor and iterate on the implementation.

**Comment by DarkLight1337** [comment]
cc @WoosukKwon

**Comment by Trenx-J** [comment]
Friendly ping @WoosukKwon.

We have considered your concern about engine‑core changes. Our implementation is designed to support a family of generative recommendation models with minimal core modifications, and we are ready to refactor further if there are code‑logic issues.

Would you mind taking another look when you have bandwidth? Thanks a lot.

**Comment by Trenx-J** [comment]
> cc @WoosukKwon

Friendly ping @DarkLight1337 .

We pinged WoosukKwon earlier with no reply. Could you help route this PR to another expert who knows the engine‑core area when you have time? We are open to further refactoring. Thanks a lot.
